src/catalog/src/durable/impls/migrate.rs

```diff
@@ -0,0 +1,227 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use async_trait::async_trait;
+use mz_ore::now::EpochMillis;
+
+use crate::durable::debug::{DebugCatalogState, Trace};
+use crate::durable::impls::persist::UnopenedPersistCatalogState;
+use crate::durable::impls::stash::OpenableConnection;
+use crate::durable::{
+    BootstrapArgs, CatalogError, DurableCatalogState, OpenableDurableCatalogState,
+};
+
+// Note: All reads done in this file can be fenced out by a new writer. All writers start by first
+// fencing the stash and then fencing persist. Therefore, when performing operations like:
+//
+//   1. Read tombstone from stash.
+//   2. Use tombstone to decide to either read something from the stash or persist.
+//
+// We don't have to worry about race conditions where a new writer updates the stash or persist
+// inbetween steps (1) and (2).
+
+#[derive(Debug)]
+pub(crate) enum Direction {
+    MigrateToPersist,
+    RollbackToStash,
+}
+
+#[derive(Debug)]
+pub struct CatalogMigrator {
+    openable_stash: Box<OpenableConnection>,
+    openable_persist: Box<UnopenedPersistCatalogState>,
+    direction: Direction,
+}
+
+#[async_trait]
+impl OpenableDurableCatalogState for CatalogMigrator {
```

Comment:
Why does this need to implement the trait?

Reply:
The idea is that during the release where we are upgrading environments, we always open the catalog with `CatalogMigrator` and never with the more specific implementations. There's probably alternate approaches, but here's some reasons why I thought this was a good approach:

1. Opening the underlying catalogs is part of the migration/rollback, so I thought it would be a good idea to do all the migration/rollback during the opening of the catalog. This allows us to leave the calling code as is, and just change the `OpenableDurableCatalogState` implementation.
2. A previous migration/rollback may have crashed half-way through and so we always want to open the catalog with an implementation that is aware of both underlying catalogs, knows how to delegate reads depending on the current state, and can finish an unfinished migration/rollback. This is especially true of the pre-open reads like `get_deployment_generation`.